packages/ai-jsx/src/lib/tts/tts.tsx

```diff
@@ -0,0 +1,360 @@
+ * generated, rather than waiting for the entire audio file to be generated. It also
+ * allows text to be fed to the service in a stream rather than all at once. This
+ * class is not meant to be used directly, but provides infrastructure for
+ * RestTextToSpeech and ElevenLabsTextToSpeech.
+ */
+export class MseTextToSpeech extends TextToSpeechBase {
+  private readonly mediaSource: MediaSource = new MediaSource();
+  private sourceBuffer?: SourceBuffer;
+  private readonly chunkBuffer: AudioChunk[] = [];
+  private inProgress = false;
+
+  constructor(name: string) {
+    super(name);
+    this.audio.src = URL.createObjectURL(this.mediaSource);
+    this.audio.onwaiting = () => {
+      console.log(`[${this.name}] tts waiting`);
+      if (!this.inProgress) {
+        this.onComplete?.();
+      }
+    };
+    this.mediaSource.onsourceopen = () => {
+      if (!this.sourceBuffer) {
+        this.sourceBuffer = this.mediaSource.addSourceBuffer('audio/mpeg');
+        this.sourceBuffer.onupdateend = () => {
+          this.processChunkBuffer();
+        };
+      }
+    };
+  }
+  play(text: string) {
+    if (this.audio.readyState == 0) {
+      console.log(`[${this.name}] tts starting play`);
+      this.audio.play();
+    }
+    if (!this.inProgress) {
+      this.playMillis = performance.now();
+      this.inProgress = true;
+    }
+    this.generate(text);
+  }
+  flush() {
+    if (this.inProgress) {
+      this.doFlush();
+    }
+  }
+  protected queueChunk(chunk: AudioChunk) {
+    this.chunkBuffer.push(chunk);
+    this.processChunkBuffer();
+  }
+  protected processChunkBuffer() {
+    if (!this.sourceBuffer?.updating && this.chunkBuffer.length > 0 && this.chunkBuffer[0].buffer) {
+      const chunk = this.chunkBuffer.shift();
+      this.sourceBuffer?.appendBuffer(chunk!.buffer!);
+    }
+  }
+  protected generate(_text: string) {}
+  protected doFlush() {}
+  protected setComplete() {
+    this.inProgress = false;
+  }
+  protected cleanUp() {
+    this.chunkBuffer.length = 0;
+    this.inProgress = false;
+    super.cleanUp();
+  }
+}
+
+/**
+ * Defines a text-to-speech service that requests chunked audio from a server
+ * using REST requests for each chunk of audio.
+ */
+export class RestTextToSpeech extends MseTextToSpeech {
+  private pendingText: string = '';
+  constructor(
+    name: string,
+    private readonly urlFunc: BuildUrlFunction,
+    private readonly voice: string,
+    private readonly rate: number = 1.0
+  ) {
+    super(name);
+  }
+  protected async generate(text: string) {
+    this.pendingText += text;
+    while (true) {
+      // Find the first punctuation mark (period, exclamation point, or question mark)
+      // that is not followed by another character (e.g., not the dot in $2.59).
+      // We'll send off the resultant sentence for generation, and jump ahead, skipping
+      // any spaces after the punctuation.
+      const index = this.pendingText.search(/[.!?][^\w]/);
```

Comment:
This is fine, but if you'd like to try something more robust: https://www.perplexity.ai/search/6a70b2c1-08bf-4b84-b494-3b07afaf3e2a?s=u